esp32/espnow: Add support for espnow v2.0.

AmirHmZz opened the pull request:
### Summary

2 years ago, [I've submitted a feature request](https://github.com/espressif/esp-idf/issues/9453) to increase maximum length of `ESP-NOW` messages. [Finally](https://github.com/espressif/esp-idf/commit/b4d102796f50e6fabce0f2a1c6ff5b8f39c902a2) the maximum length increased from **250** bytes to **1490** bytes in `ESP-NOW v2.0` which is implemented in `IDFv5.4`. This PR aims to migrate current implementation of `espnow` module to support `ESP-NOW v2.0` but also keep support for `ESP-NOW v1.0` for builds with older `IDF` versions.


### Testing

I've tested both sending and receiving examples on `ESP32` and `ESP32-S3` chips.

AmirHmZz committed:
esp32/espnow: Add support for espnow v2.0.

Co-authored-by: Angus Gratton <angus@redyak.com.au>
Signed-off-by: Amirreza Hamzavi <amirrezahamzavi2000@gmail.com>

glenn20 commented:
@AmirHmZz  Thanks for your persistence in getting this change into the espressif IDE and now to producing this PR. I will endeavour to provide an initial review in the next few days.

gretel commented:
thanks!

AmirHmZz commented:
@glenn20 still waiting for your review!

projectgus reviewed (commented):
Thanks @AmirHmZz for submitting this. I was unaware of ESPNow V2 until I went to implement a fix for #18311 (unrelated), it looks quite useful!

Unfortunately there is some additional work needed to make sure we don't break existing MicroPython code when adding V2 support (see inline comments).

There is also a set of tests in `tests/multi_espnow` (run as `./run-multitests.py -p2 -t PORTA -t PORTB multi_espnow/*.py`). These used to pass on IDF <5.4, but at least one test (`20_send_echo.py`) is already broken in master (can already send long ESPNow messages with MicroPython, but can't receive them properly). It'd also be good to add a test specifically for the V2 long messages (that we can skip when running on esp8266). Are you able to work on the tests as part of this PR as well?

projectgus commented:
> at least one test (`20_send_echo.py`) is already broken in master

Oh, FWIW there's another failure in master + IDF 5.5.1 that I think is an ESP-IDF bug, `multi_espnow/70_channels.py` fails about 25% of the time *when transmitting from some chips only*. They seem to take a very long time to switch channels (so the message is sent to the old/wrong channel). If I have time I'll try to reproduce in C code and report to ESP-IDF.

AmirHmZz commented:
@projectgus Sorry for the late response. I'm still interested in this, but I live in Iran and the internet here was completely cut off for about 40 days. It just came back recently, but only GitHub and a few other services are accessible right now, and there's no guarantee it won't go down again.

Given the instability, please feel free to make those updates cause I definitely don't want this PR to stall because of my connectivity issues.

projectgus commented:
@AmirHmZz Wow, I appreciate you getting back to me at all. You've certainly got more important things to worry about than this PR, I'll try and make some small updates to get it merged. Hoping you can stay stay safe and well.

projectgus committed:
tests/multi_espnow: Add a specific test for ESP-NOW V2 packet lengths.

Signed-off-by: Angus Gratton <angus@redyak.com.au>

projectgus committed:
esp32/espnow: Clear the recv_cb & arg on deinit.

These won't be accessed again until a new callback is set, but it stops the
objects being kept alive unnecessarily.

This work was funded through GitHub Sponsors.

Signed-off-by: Angus Gratton <angus@redyak.com.au>

projectgus commented on docs/library/espnow.rst:
```diff
@@ -148,10 +157,10 @@ Configuration
     .. data:: Options:
 
-        *rxbuf*: (default=526) Get/set the size in bytes of the internal
+        *rxbuf*: (default=528 or 3008) Get/set the size in bytes of the internal
```
Decided to keep this as-is, so updated ESP32s will get the larger buffer. Added a note in the docs about reducing this value to save RAM if only using ESP-NOW V1.

projectgus commented:
I've made some tweaks to @AmirHmZz's work and added a test.

I also noticed while working on this that Espressif reduced the maximum packet size for V2 to 1470 bytes during ESP-IDF v5.5 development: https://github.com/espressif/esp-idf/commit/a4106ef3b94ae2c90dec0ed5364b2290966c7a6e (!!!), so have updated this PR to match.

Trying to get the multi_espnow tests to pass I fixed some other test issues, which I'll submit in another PR.

dpgeorge reviewed (approved):
Thanks @projectgus for updating this PR.

I tested using IDF 5.5.2 with ESP32_GENERIC and ESP32_GENERIC_S2, and I can get all `multi_espnow` tests to pass (most of the time).

Good to see docs and tests updated/added.  This is good to merge now.
